mne/inverse_sparse/mxne_optim.py

```diff
@@ -714,26 +714,141 @@ def norm_l1_tf(Z, shape, n_orient):
+        active_set = np.tile(active_set[:, None], [1, n_orient]).ravel()
+    for o in range(n_orient):
+        Y[o::n_orient] *= shrink
+    return Y
+
+
+def dgap_l21l1(M, G, Z, active_set, alpha_space, alpha_time, phi, phiT, shape,
+               n_orient, highest_d_obj, screening):
+    """Duality gaps for the time-frequency mixed norm inverse problem
+
+    For details see:
+    Wang J., and Ye J.,
+    Two-layer feature reduction for sparse-group lasso via decomposition of
+    convex sets, in Advances in Neural Information Processing Systems (NIPS),
+    vol. 27, pp. 2132-2140, 2014.
+
+    Bach F., Jenatton R., Mairal J., and Obozinski G.,
+    Optimization with sparsity-inducing penalties, Foundations and Trends
+    in Machine Learning, vol. 4, no. 1, pp. 1-106, 2012.
+
+    Gramfort A., Kowalski M., and Hamalainen, M.,
+    Mixed-norm estimates for the M/EEG inverse problem using accelerated
+    gradient methods, Physics in Medicine and Biology, 2012
+    http://dx.doi.org/10.1088/0031-9155/57/7/1937
+
+    Parameters
+    ----------
+    M : array of shape [n_sensors, n_times]
+        The data.
+    G : array of shape [n_sensors, n_active]
+        Gain matrix a.k.a. lead field.
+    Z : array of shape [n_active, n_coefs]
+        Sources in TF domain.
+    active_set : array of bool
+        Mask of active sources.
+    alpha_space : float
+        The spatial regularization parameter. It should be between 0 and 100.
+    alpha_time : float
+        The temporal regularization parameter. The higher it is the smoother
+        will be the estimated time series.
+    phi : instance of _Phi
+        The TF operator.
+    phiT : instance of _PhiT
+        The transpose of the TF operator.
+    shape : tuple
+        Shape of TF coefficients matrix.
+    n_orient : int
+        Number of dipoles per locations (typically 1 or 3).
+    highest_d_obj : float
+        The highest value of the dual cost so far.
+    screening : array of bool
+        Mask of screened-out sources by dynamic screening.
+
+    Returns
+    -------
+    gap : float
+        Dual gap
+    p_obj : float
+        Primal cost
+    d_obj : float
+        Dual cost. gap = pobj - dobj
+    R : array of shape [n_sensors, n_times]
+        Current residual of M - G * X
+    screening : array of bool
+        Mask of screened-out sources by dynamic screening.
+    """
+    X = phiT(Z)
+    GX = np.dot(G[:, active_set], X)
+    R = M - GX
+    penaltyl1 = norm_l1_tf(Z, shape, n_orient)
+    penaltyl21 = norm_l21_tf(Z, shape, n_orient)
+    nR2 = sum_squared(R)
+    p_obj = 0.5 * nR2 + alpha_space * penaltyl21 + alpha_time * penaltyl1
+
+    GRPhi = prox_l1_full(phi(np.dot(G[:, screening].T, R)), alpha_time,
+                         n_orient)
+    GRPhi = stft_norm2(GRPhi.reshape(*shape)).reshape(-1, n_orient)
+    GRPhi = np.sqrt(GRPhi.sum(axis=1))
+    dual_norm = np.amax(GRPhi)
+    scaling = alpha_space / dual_norm
+    scaling = min(scaling, 1.0)
```

Comment:
since the computation of scaling is not trivial (and the references don't have the same notation), could we include a .rst somewhere detailing a bit the computation? We can use my gist